row_transform/row_transform.tex

```diff
@@ -0,0 +1,188 @@
+\evensidemargin0.8cm
+\textheight22.7cm 
+\textwidth15cm \topmargin-0.5cm
+
+\newtheorem{theorem}{Theorem}[section]
+\newtheorem{lemma}[theorem]{Lemma}
+
+\usepackage{listings}
+\usepackage{xcolor}
+
+\definecolor{codegreen}{rgb}{0,0.6,0}
+\definecolor{codegray}{rgb}{0.5,0.5,0.5}
+\definecolor{codepurple}{rgb}{0.58,0,0.82}
+\definecolor{backcolour}{rgb}{0.95,0.95,0.92}
+
+\lstdefinestyle{mystyle}{
+    backgroundcolor=\color{backcolour},   
+    commentstyle=\color{codegreen},
+    keywordstyle=\color{magenta},
+    numberstyle=\tiny\color{codegray},
+    stringstyle=\color{codepurple},
+    basicstyle=\ttfamily\footnotesize,
+    breakatwhitespace=false,         
+    breaklines=true,                 
+    captionpos=b,                    
+    keepspaces=true,                 
+    numbers=left,                    
+    numbersep=5pt,                  
+    showspaces=false,                
+    showstringspaces=false,
+    showtabs=false,                  
+    tabsize=2
+}
+
+\lstset{style=mystyle}
+
+\newcommand{\silvia}[1]{{ {\color{blue}{(silvia)~#1}}}}
+\newcommand{\grace}[1]{{ {\color{purple}{(grace)~#1}}}}
+
+\newcommand{\MultiSet}{\mathrm{MultiSet}}
+\newcommand{\len}{\mathrm{len}}
+\newcommand{\din}{\texttt{d\_in}}
+\newcommand{\dout}{\texttt{d\_out}}
+\newcommand{\T}{\texttt{T} }
+\newcommand{\F}{\texttt{F} }
+\newcommand{\Relation}{\texttt{Relation}}
+\newcommand{\X}{\mathcal{X}}
+\newcommand{\Y}{\mathcal{Y}}
+\newcommand{\True}{\texttt{True}}
+\newcommand{\False}{\texttt{False}}
+\newcommand{\clamp}{\texttt{clamp}}
+\newcommand{\function}{\texttt{function}}
+\newcommand{\float}{\texttt{float }}
+\newcommand{\questionc}[1]{\textcolor{red}{\textbf{Question:} #1}}
+
+\title{Privacy Proofs for OpenDP: Row Transform}
+\author{Grace Tian}
+\date{Summer 2021}
+\begin{document}
+
+\maketitle
+\tableofcontents
+
+
+\section{Algorithm Implementation}
+\subsection{Code in Rust}
+The current OpenDP library contains the \texttt{make\_row\_by\_row} function implementing the row transform function. This is defined in lines 10-26 of the file \texttt{manipulation.rs} in the Git repository (\url{https://github.com/opendp/opendp/blob/main/rust/opendp/src/trans/manipulation.rs#L10-L26}).
+
+\includegraphics[width=\textwidth]{make_row_by_row.jpg}
+
+
+\subsection{Pseudo Code in Python} \label{sec:pseudocode}
+
+\subsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{atom\_input\_domain} has type \texttt{DIA}
+        \item Variable \texttt{atom\_output\_domain} has type \texttt{DOA}
+        \item Variable \texttt{atom\_function} has type \texttt{F}
+        \item Types \texttt{DIA} and \texttt{DOA} have trait \texttt{Domain}
+        \item Type \texttt{F} has trait \texttt{Fn(\&DIA::Carrier) -> DOA::Carrier} 
+    \end{itemize}
+    \item \texttt{atom\_function} is a pure function
```

Comment:
This should say "atom_function" is a pure *randomized* function, and we need to include a definition of that here or point to one of the defns docs.  We also need atom_function to have the property that it has an "appropriate output domain" property, i.e. maps any element of DIA to an element of DOA.

Reply:
May be also worth adding a warning that some choices of atom_function can also lead to timing attacks.  Although we are not promising protection against timing attacks in this version of the library, it is still something users of make_row_by_row should be careful about.